### ISSUE TITLE
Rescope still-generic typevars in return callables

### ISSUE THREAD

**Issue by dcreager** [astral-sh/ty#3642]
This was already existing as a regression, but https://github.com/astral-sh/ruff/pull/24949 made it show up in more places and become more obvious.

A short repro (that doesn't depend on https://github.com/astral-sh/ruff/pull/24949) is the following:

```py
def higher[A, B](f: Callable[[A], B]) -> Callable[[A], B]: ...
def id[T](x: T) -> T: ...

# currently: (A@higher, /) -> A@higher
# should be: [A](A, /) -> A
reveal_type(higher(id))

# currently: [invalid-argument-type]
# should be: Literal[1]
reveal_type(higher(id)(1))

# currently: [invalid-argument-type]
# should be: Literal["x"]
reveal_type(higher(id)("x"))
```

Note that the first reveal mentions a typevar but does not bind that typevar in a generic context. That means that in the later call, we do not consider the typevar to be inferable, and do not infer (different) specializations for each of the calls.

It's correct that the return type contains typevars in its `Callable` signature. The error is that we should detect those lingering typevars and "rescope" them by creating a new generic context for the returned callable.
